oak_functions/loader/src/server.rs

```diff
@@ -149,25 +148,54 @@ impl FixedSizeBodyPadder for Response {
     }
 }
 
+/// Trait for implementing extensions, to implement new native functionality.
+pub trait OakApiNativeExtension {
+    /// Similar to `invoke_index` in [`wasmi::Externals`], but may return a result to be
+    /// written into the memory of the `WasmState`.
+    fn invoke_index(
+        &self,
+        wasm_state: &WasmState,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Result<Option<ExtensionResult>, OakStatus>, wasmi::Trap>;
+
+    /// Similar to `resolve_func` in [`wasmi::ModuleImportResolver`], but returns a tuple with a
+    /// function index and a function signature to be converted into [`wasmi::FuncRef`].
+    fn resolve_func(&self) -> anyhow::Result<(usize, wasmi::Signature)>;
```

Comment:
IIUC this only works if the extension exposes a single function; for multiple functions, the API would be more complex. In particular, for the current API, the method `invoke_index` is not appropriately named, since the index does not even appear in the function signature.